test/quantization/test_quantized_op.py

```diff
@@ -2395,6 +2395,100 @@ def test_custom_module_lstm(self):
                         msg=(f"Error is too high: SNR(dB): {power}, "
                              f"Signal: {signal}, MSE: {mse}"))
 
+    @override_qengines
+    def test_custom_module_multi_head_attention(self):
+        class MultiheadAttentionModel(torch.nn.Module):
+            def __init__(self, *args, **kwargs):
+                super(MultiheadAttentionModel, self).__init__()
+                self.layer = torch.nn.MultiheadAttention(*args, **kwargs)
+
+            def forward(self, *args, **kwargs):
+                return self.layer(*args, **kwargs)
+
+        qengine = torch.backends.quantized.engine
+
+        num_heads = 16
+        batch_size = 4
+        target_seq_length = 128
+        source_seq_length = 64
+        embed_dim = 512  # Must be divisible buy the number of heads
+
+        dropout = 0  # This is not supported
+
+        Bias = [False, True]
+        Add_bias_kv = [False, True]
+        Add_zero_attn = [False, True]
+
+        dtype = np.uint8
+        qtype = torch.quint8
+
+        custom_module_config = {
+            'float_to_observed_custom_module_class': {
+                torch.nn.MultiheadAttention: torch.nn.quantizable.MultiheadAttention
+            },
+            'observed_to_quantized_custom_module_class': {
+                torch.nn.quantizable.MultiheadAttention: torch.nn.quantizable.MultiheadAttention
+            }
+        }
+
+        fp_data = [
+            torch.randn(target_seq_length, batch_size, embed_dim),  # Q
+            torch.randn(source_seq_length, batch_size, embed_dim),  # K
+            torch.randn(source_seq_length, batch_size, embed_dim)   # V
+        ]
+
+        q_data = []
+        for idx, x in enumerate(fp_data):
+            scale, zero_point = _calculate_dynamic_qparams(x, dtype=dtype)
+            x = x.to(torch.float)
+            qx = torch.quantize_per_tensor(x, scale=scale,
+                                           zero_point=zero_point, dtype=qtype)
+            q_data.append(qx)
+
+            # Dequantize the data back for reference
+            fp_data[idx] = qx.dequantize()
+
+        with torch.no_grad():
+            for bias, add_bias_kv, add_zero_attn in itertools.product(
+                    Bias, Add_bias_kv, Add_zero_attn):
+                # Assume 12dB is sufficient for functional equivalence
+                # Without the bias, linear might perform poorly
+                min_power = 12 if bias else 10
```

Comment:
Can we do a tighter check? If you can use the dequantized versions of the quantized MHA weights, then the outputs should match very closely.

Reply:
In my tests the snr was pretty low -- I'll try the q/dq weights